// default_router/index.js
//exports.metadata = require('./metadata.json');
var url 	= require('url');
var fs		= require('fs');
var ac		= require('ac');


var configuration = {
	'ac'		: 	ac,
	'caching'	: 	true,
};
exports.configuration = configuration;


var handlers = {
	request : { 
		'GET' 		: [], 
		'POST'		: [],
		'DELETE'	: [] 
	},
	statusCode : {
		404 : function(){}
	}
};

exports.get = function(condition, callback, extended_options){
	if(typeof condition == 'undefined' || typeof callback == 'undefined')
		return handlers.request.GET[0];
	handlers.request.GET.push({'condition' : condition, 'callback':callback});
};

exports.post = function(condition, callback, extended_options){
	if(typeof condition == 'undefined' || typeof callback == 'undefined')
		handlers.request.POST[0];
	handlers.request.POST.push({'condition' : condition, 'callback':callback});
};

exports.status = function(code, callback, extended_options){
	if(typeof callback== 'undefined'){
		return handlers.statusCode[code];
	}
	handlers.statusCode[code] = callback;
};

exports.requestHandler = function(request, response){
	var handlerArray = handlers.request[request.method];
	if(typeof handlerArray == 'undefined'){

	}
	else{
		var parsedURL = url.parse(request.url);
		for(var Index = handlerArray.length-1; Index >= 0; Index--){
			var handler = handlerArray[Index];
			if(handler.condition.constructor.name == "String"){
				if(handler.condition != parsedURL.pathname){
					continue;
				}
				handler.callback(request, response);
				return;
			}
			else if(handler.condition.constructor.name == "RegExp"){
				var matches = parsedURL.pathname.match(handler.condition);
				if(typeof matches != 'undefined'){
					continue;
				}
				request.params = matches;
				handler.callback(request, response);
				return;
			}else if(typeof handler.condition == 'function'){
				if(!handler.condition(request, parsedURL.pathname)){
					continue;
				}
				handler.callback(request, response);
				return;
			}
		}
	}
};
//TODO: Add use of ac (access control) to...well...access....control

exports.get(/^(.*)\/$/, function(request, response){
	var content = fs.readFileSync('./' + request.params[0] + 'index.html');
	if(typeof content == 'undefined'){
		if(typeof handlers.statusCode[404] != 'undefined' && !handlers.statusCode[404](request, response)){
			response.writeHead(404, {"Content-Type": "text/html"});
			response.write("404! no index found!");
			response.end();
			return;
		}
	}
	response.writeHead(200, {"Content-Type": "text/html"});
	response.write(content)
	response.end();
	return;
});
